Set up Postgres Database

oz03-hub opened the pull request:
This PR achieves:
- Integrates a Docker service of postgres into existing codebase
- Updates readme instructions on how to run and develop with the database
- Updates docker-compose to use correct .env files and pass secrets about database into containers
- Has initial api routes to check connection to database from backend
- Has a separate directory `backend/app/db/` for db related scripts, use this directory to add any db related stuff, then reference it in main.py for api routes
- Updated requirements.txt to include sqlalchemy and psycopg2
- Updated gitignore to exclude .env files from being pushed to repo, message me on slack to get an .env copy

oz03-hub commented:
Look at #inaturalist channel on slack for .envs, please place them in root of your project

SergePogorelov commented on docker-compose.yml:
```diff
@@ -1,4 +1,15 @@
 services:
+  db:
+    image: postgres
```
please indicate the version / tag of the image we are using

oz03-hub replied:
16

oz03-hub commented:
You all should have gotten updated .env files following latest commit, I won't touch code from now on until people get a chance to review. @SergePogorelov your review will be dismissed, you can re-review it if you want, everything is still the same but I decided to keep default postgres port

oz03-hub commented:
@cheerstopaula your changes were added in the new commit.
Changes include:
- everything in paula/database branch
- re-organizing src/backend, bundled database related files into backend/app/db
- removed parts of code that 'cached' predictions and tried to load it
- removed code that serves previous annotation for a species while generating predictions, now with each generation you start with a clean state

cheerstopaula commented on README.md:
```diff
@@ -27,6 +27,20 @@ Note: `src/backend/sinr` is a submodule from `UMassCDS/inatrualist-sinr`, if you
 
+## Working with database
+
+1. Notice that there are two .env files, .docker.env and .env, .env is used for your local development, .docker.env will be used in your containers.
```
This can be less redundant. Maybe "Notice that there are two .env files -- used for your local development-- and .docker.env -- used for your containers/"

This shouldn't be listed as 1., since you don't really need to take action. Also might be good to explain why the containers AND local are needed at the same time.

cheerstopaula commented on README.md:
```diff
@@ -27,6 +27,20 @@ Note: `src/backend/sinr` is a submodule from `UMassCDS/inatrualist-sinr`, if you
+
+1. Notice that there are two .env files, .docker.env and .env, .env is used for your local development, .docker.env will be used in your containers.
+
+2. To run the application without containers, first run the database container by `docker compose up db` (you may have to rebuild it if you changed docker-compose.yml). Run `uvicorn src.backend.app.main:app --reload --env-file .env` at project root to run the server, then `npm start` in `src/frontend`.
```
There is a docker section in the readme, this section should go first or maybe be referenced here. Users might be blindsided when looking at the docker command

cheerstopaula replied:
Docker compose up db should be one step. 
Open a different terminal, run uvicorn should be another step. 
Open a third terminal, run npm start should be another step.

cheerstopaula replied:
To run uvicorn and npm, we first need to set up the Conda environment and run npm I, why are these steps after? they should be higher up!

cheerstopaula commented on README.md:
````diff
@@ -56,13 +58,20 @@ npm i --prefix src/frontend/
 2. Launch the **backend**:
 ```bash
- uvicorn src.backend.app.main:app --reload
+ uvicorn src.backend.app.main:app --reload --env-file .env
````
To run uvicorn and npm, we first need to set up the Conda environment and run npm I, why are these steps after? they should be higher up!

cheerstopaula commented on README.md:
````diff
@@ -56,13 +58,20 @@ npm i --prefix src/frontend/
+```
+
+## Run **backend** in second terminal:
 1. Navigate to the main `ds4cg2024-inaturalist` directory if you are not already there:
````
Conda environment needs to be activated here, add to instructions.

angelazhu2 commented on README.md:
```diff
@@ -17,18 +17,20 @@ Estimating the geographical range of a species from sparse observations is a cha
+
+4. You are all set for setting up the code.
+
+Note: We need .env for local development, and .docker.env for docker containers. The difference between them is database url, which is caused by how docker manages networks.
```
consolidate all the notes into one big Note. or just "additional information" and have them be bullet points with bolding for the main point of each thing.

angelazhu2 replied:
@oz03-hub  can you consolidate these notes into one with just bullet point.

angelazhu2 commented on README.md:
````diff
@@ -56,13 +58,20 @@ npm i --prefix src/frontend/
 2. Launch the **backend**:
 ```bash
- uvicorn src.backend.app.main:app --reload
+ uvicorn src.backend.app.main:app --reload --env-file .env
````
note if you run into errors, try:
- pip installing requriements.txt (with ur command)
- installing postgresql (i used homebrew: brew install postgresql)

oz03-hub replied:
psycopg2-binary might be a solution

oz03-hub replied:
but I will add brew i postgres as a troubleshooting alternative

cheerstopaula commented on README.md:
```diff
@@ -87,6 +96,22 @@ Note: you don't have to initialize submodule to run docker, dockerfile will set
+
+1. You can run the application without containers, allowing you to develop things quickly. Refer to **Running the iNatAtor Application** above for instructions.
+
+2. You can verify connections are working by going to `localhost:8000/health` and `localhost:8000/echo_env`. Note that these paths are included right now to assist with rapid development, will be removed for production.
```
1 and 2 are not actions, are just notes. I suggest keeping number only on actions.

oz03-hub replied:
I will reword those, I think it's still important to verify connection as a step

oz03-hub commented on .gitignore:
```diff
@@ -103,6 +103,7 @@ celerybeat.pid
 
 # Environments
 .env
+.docker.env
```
in instructions these are told to be given in private

ginic replied:
Yeah, it's a good idea to have them in the gitignore. You don't want to accidentally commit them to a public repo, because they often will contain private keys or other sensitive info. So long as the README lists the variables that need to be in .env or otherwise set, that's good!

oz03-hub commented:
Thanks for the reviews, I made changes to readme as requested, I tried to break up some instructions and reword some instructions into just paragraphs because they weren't really steps. Also replaced psycopg2 with binary package which should contain necessary postgres stuff

SergePogorelov commented on src/backend/app/db/database.py:
```diff
@@ -0,0 +1,10 @@
+from sqlalchemy.ext.declarative import declarative_base
+import os
+
+DB_URL = os.getenv("DATABASE_URL")
```
here instead of getting the URL get the DEBUG and associated with it create a corresponding URL for the DB

SergePogorelov commented on src/backend/app/db/models.py:
```diff
@@ -0,0 +1,15 @@
+
+class Annotation(Base):
+    __tablename__="Annotation"
+    AnnotationID = Column(Integer, primary_key=True, index=True)
```
usually for the PostgreSQL and other SQL DB people using **lower_case_with_underscores**
https://stackoverflow.com/questions/2878248/postgresql-naming-conventions

and this also matches the PEP8 style in python

oz03-hub replied:
I wish this was pointed out earlier in the database proposal stage, I still think it should be addressed to the proposal and not this task as this task follows whatever the proposal introduced

oz03-hub replied:
there is not really a standard I found, that being said as long as we are consistent with whatever way we choose I don't see a problem

SergePogorelov replied:
maybe it not a standart for SQL (never never seen field names like that though), but it still doesn't fit PEP8 :)

SergePogorelov commented on src/backend/app/main.py:
```diff
@@ -17,13 +29,41 @@
     allow_headers=["*"],
 )
 
+@app.get("/health")
```
> if now the database is already connected to the project and saves annotations, then we no longer need these functions

SergePogorelov commented on src/backend/app/main.py:
```diff
@@ -17,13 +29,41 @@
 
 @app.post("/save_annotation/")
-async def save_annotation(request: Request):
-    response = tools.save_annotation(await request.json())
```
if we no longer use some functions, then they can be removed (in separate commits, of course)

SergePogorelov commented on src/backend/app/db/crud.py:
```diff
@@ -0,0 +1,18 @@
+
+DATABASE_URL = os.getenv("DATABASE_URL")
+
+engine = create_engine(DATABASE_URL)
```
Why do we create a database connection in two different files? It seems like some of this is superfluous...

oz03-hub replied:
right

SergePogorelov reviewed (changes requested):
I left a few comments, please take a look.

oz03-hub commented on README.md:
```diff
@@ -17,32 +17,20 @@ Estimating the geographical range of a species from sparse observations is a cha
 1. Clone the repository `git clone git@github.com:UMassCDS/ds4cg2024-inaturalist.git`
 
-2. Navigate to cloned project's root, run `git submodule init` and then `git submodule update --remote --merge`
+2. Create .env and .docker.env at the project root. Copy the contents that were given to you on **slack** into these files.
```
Good idea, there will be .env.copy that they can rename and copy contents in

oz03-hub commented on README.md:
```diff
@@ -96,17 +90,31 @@ In your web browser, open the link [http://localhost:3000/](http://localhost:300
+
+Note: The codebase is getting bigger, therefore add database related code in `src/backend/app/db`, then make proper API routes in main, if it gets too big, we can resort to using API routers from fastapi.
+
+Note: There are two environment files (.env and .docker.env) because the database url for local development and docker environments are separate.
```
I think this will introduce a lot of overhead and complication, sqlite might be good for quick development and we used to implement things quickly but we decided that postgres would be what we use in any case. @ginic your thoughts on this?

oz03-hub replied:
Running postgres is not that difficult on docker, and this request is kind of expanding the scope of the ticket

SergePogorelov replied:
I still don't like the idea of ​​running a postgresql container for local development.
if you just add 2 lines of code to database.py you can:
1. don't start docker at all
2. do not create any .env files locally

Perhaps if we decide to save predictions in the form of coordinates, then sklight will not suit us. but we don't do that yet...

```python
DB_URL = os.getenv("DATABASE_URL")
if not DB_URL:
    DB_URL = "sqlite:///./src/backend/db.sqlite3"
```

SergePogorelov replied:
yes, it should work. But now we have to not forget to create  `.env`  file to put this link there)

ginic replied:
Sounds good! I will add a suggestion so that information is in the README

oz03-hub commented on README.md:
```diff
@@ -96,17 +90,31 @@ In your web browser, open the link [http://localhost:3000/](http://localhost:300
+
+Sometimes you want to run the application without containers, allowing you to develop things quickly. The **Running the iNatAtor Application** section explains how to run the application locally.
+
+You can verify connections are working by going to `localhost:8000/health` and `localhost:8000/echo_env`. Note that these paths are included right now to assist with rapid development, will be removed for production.
```
I still would like to keep /health/ because some people don't know how to communicate with postgres through terminal or don't know how to check if things are running correctly on docker, /health/ is an easy way to make sure thing is up and running. echo_env can be removed

ginic commented on docker-compose.yml:
```diff
@@ -1,4 +1,14 @@
 services:
+  db:
```
It might be good to give this a `container_name` value as well, so it doesn't just get named 'postgres'. This will make debugging easier. I'd recommend `inatator_db` or something similar.

ginic reviewed (changes requested):
I tested this with both the local set up and the docker compose up version. Was able to save and verify my annotations are saved. Everything is working great and the schema matches the proposal for the demo!

I really like the way you are handling the database connection Sessions with Fast API's Depends, it makes the code clean and easy to follow while making sure the database session is used efficiently. 

I have a couple mostly minor suggestions in comments. 
The one big thing that needs to be changed is to always make sure that you're using a `with` statement when working with DB connections. They're a little less error prone in python than a `try/finally`.

oz03-hub commented on src/backend/requirements.txt:
```diff
@@ -7,3 +7,5 @@ torch==1.12.1
 uvicorn==0.30.1
 alphashape==1.3.1
+sqlalchemy
+psycopg2-binary
```
yes, it was originally psycopg2, the binary version seemed to include things needed from postgres as well, we switched because some people in the team didn't have postgres installed and were having a hard time setting up

oz03-hub commented:
Thanks for reviews, new commit has implemented changes (except sqlite stuff) from sergei and virginia. Please re-review one you is enough for approval

SergePogorelov commented on src/backend/app/db/database.py:
```diff
@@ -0,0 +1,14 @@
+import os
+
+DB_URL = os.getenv("DATABASE_URL")
+
```
```suggestion
if not DB_URL:
    DB_URL = "sqlite:///./src/backend/db.sqlite3"
   
```

oz03-hub commented:
Changed table structure, remove the postgres_data directory at the root of your project, then run `docker compose up --build db`, then uvicorn and then npm start commands you know to run the app, it works the same

ginic commented on README.md:
```diff
@@ -17,18 +17,24 @@ Estimating the geographical range of a species from sparse observations is a cha
 1. Clone the repository `git clone git@github.com:UMassCDS/ds4cg2024-inaturalist.git`
 
-2. Navigate to cloned project's root, run `git submodule init` and then `git submodule update --remote --merge`
+2. You are given .env.copy and .docker.env.copy, rename these files to .env and .docker.env respectively. Set the following environment variables in each of these files for your desired database configuration: 
+    - `POSTGRES_DB`: Name of your database
+    - `POSTGRES_USER`: Username the server will use to connect to the database
+    - `POSTGRES_PASSWORD`: Password the server will use to connect to the database
+    - `DATABASE_URL`: SQLAlchemy database connection URL. This should be something like `postgresql://${POSTGRES_USER}:${POSTGRES_PASSWORD}@db:PORT_NUMBER/${POSTGRES_DB}`, but refer to the [SQLAlchemy Database Engine docs](https://docs.sqlalchemy.org/en/20/core/engines.html#database-urls) for more details.
```
Added a note about the option to use SQLite as well: 
```suggestion
2. For local development and testing, you can choose a database engine from two options, PostgreSQL or SQLite. The most important thing is to ensure the `DATABASE_URL` is configured appropriately for your database according to [SQLAlchemy Database Engine docs](https://docs.sqlalchemy.org/en/20/core/engines.html#database-urls). To aid in setting up the database we've provided example environment files, .env.copy and .docker.env.copy, where the environment variables are listed. You can copy them to .env or .docker.env and fill in the values. 

    a. _PostgreSQL_: Run Postgres in a [local Postgres server](https://www.postgresql.org/docs/current/server-start.html) or Docker container (see provided docker-compose.yml). You should configure the following  environment variables in your environment file (.env or .docker.env).
    - `POSTGRES_DB`: Name of your database
    - `POSTGRES_USER`: Username the server will use to connect to the database
    - `POSTGRES_PASSWORD`: Password the server will use to connect to the database
    - `DATABASE_URL`: SQLAlchemy database connection URL. This should be something like `postgresql://${POSTGRES_USER}:${POSTGRES_PASSWORD}@db:PORT_NUMBER/${POSTGRES_DB}`, but 
    
    b. _SQLite_: Storing the databases in a simple SQLite database file is useful for development and testing, but shouldn't be used in production. You only need to configure the `DATABASE_URL=sqlite://<path to desired database file>`
```

ginic reviewed (approved):
Thanks, @oz03-hub ! Looks great to me! I did add a suggestion for the readme about using SQLite, but I don't think any code changes are needed.

oz03-hub commented:
Readme has sqlite option, @SergePogorelov you can configure .env in the way you like to set it up
